---
title: "R_assignment"
output: html_document
---

# PART 1

# 0. Load tidyverse
```{r}
if (!require("tidyverse")) install.packages("tidyverse")
library(tidyverse)
library(tidyr)
```

# 1. Import fang_et_al_genotypes and snp_position and analyze contents.
```{r}
genotypes <- read_tsv("./fang_et_al_genotypes.txt") 

snp_positions <- read_tsv("./snp_position.txt")

#file size
cat("\nFile Size\n")
paste0("fang_et_al_genotypes.txt is ",toString(object.size(genotypes))," bytes.")
paste0("snp_position.txt is ",toString(object.size(snp_positions))," bytes.")

cat("\nNumber of Columns\n")
paste0("fang_et_al_genotypes.txt has ", toString(ncol(genotypes))," columns.")
paste0("snp_position.txt has ", toString(ncol(snp_positions))," columns.")

cat("\nNumber of Rows\n")
paste0("fang_et_al_genotypes.txt has ", toString(nrow(genotypes))," rows.")
paste0("snp_position.txt has ", toString(nrow(snp_positions))," rows.")

```

# 2. Take columns 1,3,4 from snp_position and save as new variable.
```{r}
snp_134 <- snp_positions[c(1,3,4)]
rm(snp_positions)
```

# 3. Separate out data from Teosinte and Maize, respectively.
```{r}
#Teosinte
genotypes[c(1:3)]
Teo_genotypes <- filter(genotypes, `Group` %in% c("ZMPBA","ZMPIL","ZMPJA"))

#Maize
Maize_genotypes <- filter(genotypes, `Group` %in% c("ZMMIL","ZMMLR"," ZMMMR"))
rm(genotypes)

```

# 4. Transpose gene files created above, rename 'sample_ID' to 'SNP_ID'
```{r}
#Teosinte
t_Teosinte <- as.data.frame(t(Teo_genotypes))

#change SNP row names to its own column
t_Teosinte <- rownames_to_column(t_Teosinte,"SNP_ID")
print(t_Teosinte[1:3,])
#Remove unnecessary rows
t_Teosinte <- t_Teosinte[-(1:3),]


#Maize
t_Maize <- as.data.frame(t(Maize_genotypes))

#change SNP row names to its own column
t_Maize <- rownames_to_column(t_Maize,"SNP_ID")
print(t_Maize[1:3,])
#Remove unnecessary rows
t_Maize <- t_Maize[-(1:3),]
print(t_Maize[1:3,])
```

# 5. Arrange all three files by SNP name.
```{r}
#SNP positions
snp_sorted <- arrange(snp_134, `SNP_ID`)
rm(snp_134)

#Teosinte
Teo_sorted <- arrange(t_Teosinte,`SNP_ID`)
rm(t_Teosinte)
#Maize
Maize_sorted <- arrange(t_Maize,`SNP_ID`)
rm(t_Maize)
```

# 6. Merge teosinte and maize files each with SNP file. Create new file for each where the sign for missing data is changed from ?/? to -/-
```{r}
#Teosinte
Teo_merge_q <- merge(snp_sorted,Teo_sorted, all=TRUE)
Teo_merge_q[1:5]
Teo_merge_dash <- as.data.frame(lapply(Teo_merge_q, gsub, pattern="\\?", replacement="-"))
Teo_merge_dash[1:5]
rm(Teo_sorted)

#Maize
Maize_merge_q <- merge(snp_sorted,Maize_sorted, all=TRUE)
Maize_merge_q[1:5]

Maize_merge_dash <- as.data.frame(lapply(Maize_merge_q, gsub, pattern="\\?", replacement="-"))
Maize_merge_dash[1:5]
rm(snp_sorted)
rm(Maize_sorted)
```

# 7. Separate merge_q files' chromosomes into folders and arrange.
```{r}
#Teosinte
for (i in 1:10){
  filename = paste0('Teosinte_increasing_',i,'.txt')
  #filter by chromosome number
   x = filter(Teo_merge_q, `Chromosome` == i)
  #convert position values from characters to numbers
  x$Position <- as.numeric(x$Position)
  #order by position
   x_arr = arrange(x,`Position`)
  write.table(x_arr, file = filename,sep = '\t',quote=FALSE, row.names=FALSE)
}
test_incr <- read_tsv("Teosinte_increasing_1.txt")
rm (test_incr)
#Maize
for (i in 1:10){
  filename = paste0('Maize_increasing_',i,'.txt')
  #filter by chromosome number
  x = filter(Maize_merge_q, `Chromosome`== i)
  #convert position values from characters to numbers
  x$Position <- as.numeric(x$Position)
  #order by position
  x_arr = arrange(x,`Position`)
  write.table(x_arr, file = filename,sep = '\t',quote=FALSE, row.names=FALSE)
}
read_tsv("Maize_increasing_1.txt")

```

# 8. Separate merge_dash files' chromosomes into folders and arrange.
```{r}
#Teosinte
#having trouble getting the descending part to work. if it crashes your RStudio I'm really sorry
for (i in 1:10){
  filename = paste0('Teosinte_decreasing_',i,'.txt')
  #filter by chromosome number
  x = filter(Teo_merge_dash, `Chromosome` == i)
  #convert position values from characters to numbers
  x$Position <- as.numeric(x$Position)
  #order by position
  x_arr = arrange(x,desc(`Position))
  write.table(x_arr, file = filename,sep = '\t',quote=FALSE, row.names=FALSE)}
#test_decr <- read_tsv("Teosinte_decreasing_1.txt")
# Position keeps changing to lower numbers for some reason and i can't figure out why.
rm(test_decr)

#Maize
#for (i in 1:10){
  #filename = paste0('Maize_decreasing_',i,'.txt')
  #filter by chromosome number
  #x = filter(Maize_merge_dash, `Chromosome` == i)
  #convert position values from characters to numbers
  #x$Position <- as.numeric(x$Position)
  #order by position
  #x_arr = arrange(x,`Position`)
  #write.table(x_arr, file = filename,sep = '\t',quote=FALSE,row.names=FALSE)
#}
```



# PART 2

# 1. SNPs per Chromosome
# Plot total number of SNPs in our dataset on each chromosome
```{r}
rm(Teo_merge_dash)
rm(Maize_merge_dash)
#Teosinte
snp_count_teo <- ggplot(data = Teo_merge_q) +
  geom_bar(mapping = aes(x = Chromosome))
snp_count_teo <- snp_count_teo + ylab("SNP Count") + xlab("Teosinte Chromosome") + scale_x_discrete(labels=c("1","10","2","3","4","5","6","7","8","9","Mult.","Unk")) 
#+ scale_x_discrete(limits=c("1","2","3","4","5","6","7","8","9","10"))
print(snp_count_teo)


#Maize
snp_count_maize <- ggplot(data = Maize_merge_q) +
  geom_bar(mapping = aes(x = Chromosome))
snp_count_maize <- snp_count_maize + ylab("SNP Count") + xlab("Maize Chromosome") + scale_x_discrete(labels=c("1","10","2","3","4","5","6","7","8","9","Mult.","Unk")) 
#+ scale_x_discrete(limits=c("1","2","3","4","5","6","7","8","9","10"))
print(snp_count_maize)

```


# Plot distribution of SNPs on Chromosomes
```{r}
#Teosinte
Teo_merge_q$Position <- as.numeric(Teo_merge_q$Position)
SNP_dens_Teo <- ggplot(data = Teo_merge_q) + geom_density(mapping = aes(x=Position, fill=Chromosome),alpha =0.2)
print(SNP_dens_Teo)

#Maize
Maize_merge_q$Position <- as.numeric(Maize_merge_q$Position)
SNP_dens_Maize <- ggplot(data = Maize_merge_q) + geom_density(mapping = aes(x=Position, fill=Chromosome),alpha =0.2)
print(SNP_dens_Maize)
```

# 2. Heterozygosity
# I have a feeling this is one of the parts that requires reshaping but I"m really confused by that still so I haven't been able to figure it out :C
```{r}

```

